src/lsp4clj/server.clj

```diff
@@ -96,12 +87,17 @@
   Sends `$/cancelRequest` only once, though `lsp4clj.server/deref-or-cancel` or
   `future-cancel` can be called multiple times."
   [id method started server]
-  (map->PendingRequest {:p (promise)
-                        :cancelled? (atom false)
-                        :id id
-                        :method method
-                        :started started
-                        :server server}))
+  ;; Chaining `(-> (p/deferred) (p/catch ...))` seems like it should work, but
+  ;; doesn't.
+  (let [p (p/deferred)]
+    (p/catch p CancellationException
+      (fn [ex]
+        (protocols.endpoint/send-notification server "$/cancelRequest" {:id id})
+        (p/rejected ex)))
```

Comment:
We "fork" `p` to handle cancellation. The promise returned by `p/catch` will never be used. We can therefore return whatever, and can remove the `(p/rejected ex)` form.